Move attribute naming docs from specification (#493)
open-telemetry/semantic-conventions: docs/general/attribute-requirement-level.md

Author: Per convo with @chalin, I've formatted these files using:
```
npm run _check:format:any -- --write docs/general/attribute-*.md --prose-wrap=always
```

> # Attribute Requirement Levels for Semantic Conventions

Reviewer: ```suggestion
# Attribute Requirement Levels
```

**Status**: [Stable](../document-status.md)

<details>
<summary>Table of Contents</summary>

<!-- toc -->

- [Required](#required)
- [Conditionally Required](#conditionally-required)
- [Recommended](#recommended)
- [Opt-In](#opt-in)
- [Performance suggestions](#performance-suggestions)

<!-- tocstop -->

</details>

_This section applies to Log, Metric, Resource, and Span, and describes requirement levels for attributes defined in semantic conventions._

> Attribute requirement levels apply to the [instrumentation library](../glossary.md#instrumentation-library).

Reviewer 2: Now we will need to use the full link to the spec as we do elsewhere in the semconv repo. (referring to the spec version)

The following attribute requirement levels are specified:

- [Required](#required)
- [Conditionally Required](#conditionally-required)
- [Recommended](#recommended)
- [Opt-In](#opt-in)

The requirement level for an attribute is specified by semantic conventions depending on attribute availability across instrumented entities, performance, security, and other factors. When specifying requirement levels, a semantic convention MUST take into account signal-specific requirements.

For example, Metric attributes that may have high cardinality can only be defined with `Opt-In` level.

A semantic convention that refers to an attribute from another semantic convention MAY modify the requirement level within its own scope. Otherwise, requirement level from the referred semantic convention applies.

<!-- TODO(jsuereth) - make examples not break on changes to semconv -->

> For example, [Database semantic convention](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/README.md) references `network.transport` attribute defined in [General attributes](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/README.md) with `Conditionally Required` level on it.

Reviewer: ```suggestion
For example, [Database semantic convention](../database/README.md) references `network.transport` attribute defined in [General attributes](./README.md) with `Conditionally Required` level on it.
```

## Conditionally Required

All instrumentations MUST populate the attribute when the given condition is satisfied. The semantic convention of a `Conditionally Required` attribute MUST clarify the condition under which the attribute is to be populated.

`http.route` is an example of a conditionally required attribute that is populated when the instrumented HTTP framework provides route information for the instrumented request. Some low-level HTTP server implementations do not support routing and corresponding instrumentations can't populate the attribute.

When a `Conditionally Required` attribute's condition is not satisfied, and there is no requirement to populate the attribute, semantic conventions MAY provide special instructions on how to handle it. If no instructions are given and if instrumentation can populate the attribute, instrumentation SHOULD use the `Opt-In` requirement level on the attribute.

<!-- TODO(jsuereth) - make examples not break on changes to semconv -->

> For example, `server.address` is `Conditionally Required` by the [Database convention](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/README.md) when available. When `network.peer.address` is available instead, instrumentation can do a DNS lookup, cache and populate `server.address`, but only if the user explicitly enables the instrumentation to do so, considering the performance issues that DNS lookups introduce.

Reviewer: ```suggestion
For example, `server.address` is `Conditionally Required` by the [Database convention](../database/README.md) when available. When `network.peer.address` is available instead, instrumentation can do a DNS lookup, cache and populate `server.address`, but only if the user explicitly enables the instrumentation to do so, considering the performance issues that DNS lookups introduce.
```

> # Attribute Requirement Levels

Reviewer 3: We'll need to update all our internal links to the SPECIFICATION to this local document before you can delete the specification side, FYI.

That can be in this PR or a follow up.

Author: I've pushed a commit which updates all the references to the spec version of these documents I could find 👍

**Status**: [Stable][DocumentStatus]

<details>
<summary>Table of Contents</summary>

<!-- toc -->

- [Required](#required)
- [Conditionally Required](#conditionally-required)
- [Recommended](#recommended)
- [Opt-In](#opt-in)
- [Performance suggestions](#performance-suggestions)

<!-- tocstop -->

</details>

_This section applies to Log, Metric, Resource, and Span, and describes requirement levels for attributes defined in semantic conventions._

> Attribute requirement levels apply to the [instrumentation library](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/glossary.md#instrumentation-library).

Reviewer 2: ```suggestion
Attribute requirement levels apply to the [instrumentation library](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.26.0/specification/glossary.md#instrumentation-library)).
```
